### PR TITLE
feat(pairing): add 2FA support to pairing

### LINKED ISSUE #1185: [fxa-content-server] Support pairing for 2FA accounts
Currently you would get an error that "Accounts with 2FA cannot pair"

### PR DESCRIPTION
**Pull request by vladikoff**
Fixes #1185

### CODE REVIEW

File: packages/fxa-content-server/app/scripts/views/pair/auth_totp.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import AuthErrors from 'lib/auth-errors';
+import Cocktail from 'cocktail';
+import FormView from '../form';
+import ServiceMixin from '../mixins/service-mixin';
+import Template from '../../templates/sign_in_totp_code.mustache';
```
Comment (vladikoff) [inline packages/fxa-content-server/app/scripts/views/pair/auth_totp.js]: TODO: make sure template doesn't show `use different account` buttons

File: packages/fxa-content-server/app/scripts/views/pair/auth_totp.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    return this.model.get('account') || this.getSignedInAccount();
+  },
+
+  submit () {
+    const account = this.getAccount();
+    const code = this.getElementValue('input.totp-code');
+    return account.verifyTotpCode(code, 'pair')
+      .then((result) => {
+        if (result.success) {
+          return this.replaceCurrentPage('/pair/auth/allow', {totpComplete: true});
```
Comment (vladikoff) [inline packages/fxa-content-server/app/scripts/views/pair/auth_totp.js]: @rfk thoughts on this? 

proposal is:

* we always do TOTP for pairing if account has 2fa (verifyTotpCode)
* when `totpComplete` transitions, one check to add is when the `allow` page is `if TOTP is enabled on the account AND /totp/exists reports true` then let it through. (This is not in this PR just yet)

Comment (rfk) [reply packages/fxa-content-server/app/scripts/views/pair/auth_totp.js]: Seems reasonable to me

File: packages/fxa-content-server/tests/functional/pairing.js [hunk 3]
```diff
@@ -63,6 +66,14 @@ function getQrData(buffer) {
   });
 }
 
+const generateCode = (secret) => {
+  secret = secret.replace(/[- ]*/g, '');
+  const authenticator = new otplib.authenticator.Authenticator();
+  authenticator.options = otplib.authenticator.options;
+  const code = authenticator.generate(secret);
+  return code;
+};
```
Comment (rfk) [inline packages/fxa-content-server/tests/functional/pairing.js]: I'm surprised we don't have a shared helper for this from testing the existing TOTP flows, worth extracting into one?

File: packages/fxa-content-server/app/scripts/views/pair/pairing-totp-mixin.js [hunk 4]
```diff
@@ -9,22 +9,34 @@ export default function () {
...
       }
 
       return account.accountProfile().then((profile) => {
-        // pairing is disabled for accounts with 2FA
-        if (this.hasTotpEnabledOnAccount(profile)) {
-          this.replaceCurrentPage('pair/failure', {
-            error: AuthErrors.toError('TOTP_PAIRING_NOT_SUPPORTED'),
+        if (this.model.get('totpComplete')) {
+          // once the user with TOTP successfully verifies we check the /exists endpoint
+          // it can throw on unverified sessions or tell us if TOTP doesn't exist
```
Comment (vladikoff) [inline packages/fxa-content-server/app/scripts/views/pair/pairing-totp-mixin.js]: @rfk here's what we got

**Comment by vladikoff** [comment]
@rfk or @vbudhram please take a look

**Comment by vladikoff** [comment]
Follow up filed https://github.com/mozilla/fxa/issues/1286

File: packages/fxa-content-server/app/scripts/lib/auth-errors.js [hunk 5]
```diff
@@ -540,10 +540,10 @@ var ERRORS = {
...
     message: t('Please enter verification code')
   },
-  TOTP_PAIRING_NOT_SUPPORTED: {
-    errno: 1061,
-    message: t('Accounts with two-step authentication do not support pairing at this time')
-  },
+  // TOTP_PAIRING_NOT_SUPPORTED: {
+  //   errno: 1061,
+  //   message: t('Accounts with two-step authentication do not support pairing at this time')
+  // },
```
Comment (rfk) [inline packages/fxa-content-server/app/scripts/lib/auth-errors.js]: Is it worth adding a comment here to explain why it's now commented out?

File: packages/fxa-content-server/app/scripts/templates/sign_in_totp_code.mustache [hunk 6]
```diff
@@ -28,9 +28,12 @@
 
     </form>
 
+    {{^hideTotpAlternatives}}
     <div class="links centered">
       <a id="use-recovery-code-link" href="/signin_recovery_code">{{#t}}Use recovery code{{/t}}</a>
       <a class="different-account-link" href="/signin">{{#t}}Use a different account{{/t}}</a>
     </div>
+    {{/hideTotpAlternatives}}
```
Comment (rfk) [inline packages/fxa-content-server/app/scripts/templates/sign_in_totp_code.mustache]: Do we need another followup bug to allow the use of recovery codes in the pairing flow?

File: packages/fxa-content-server/app/scripts/views/pair/pairing-totp-mixin.js (hunk 4 above)
Comment (rfk) [reply packages/fxa-content-server/app/scripts/views/pair/pairing-totp-mixin.js]: I find the logic here a little hard to follow - if `totpComplete` has been set, doesn't that mean that we already know the user has a TOTP token and they're good to go? Why do we need to `checkTotpTokenExists()` here? Are we worried that the previous page may have set `totpComplete` incorrectly? Is it just to double-check that the session is verified? (It should definitely be verified if we just came back from the TOTP page).

File: packages/fxa-content-server/app/tests/spec/views/pair/auth_allow.js [hunk 7]
```diff
@@ -116,7 +123,56 @@ describe('views/pair/auth_allow', () => {
...
+          assert.equal(view.$el.find('.family-os').text(), 'Firefox on Windows');
+          assert.equal(view.$el.find('.location').text().trim(), 'Toronto, Ontario, Canada (estimated)');
+          assert.equal(view.$el.find('.ip-address').text(), 'IP address: 1.1.1.1');
+        });
+    });
+
+    it('handles users after TOTP with existing token throwing an error', () => {
+      model.set('totpComplete', true);
+      sinon.stub(account, 'checkTotpTokenExists').callsFake(() => {
+        return Promise.reject(new Error('Unverified Session'));
```
Comment (rfk) [inline packages/fxa-content-server/app/tests/spec/views/pair/auth_allow.js]: Hrm, so I guess this is the case that you're trying to handle that I commented on above, throwing "unverified session" after completing TOTP.  This seems quite concerning - how did we manage to complete TOTP without the session becoming verified? Have you observed this happening in the wild, after fixing the "we were using the wrong sessionToken" problem?

File: packages/fxa-content-server/app/scripts/templates/sign_in_totp_code.mustache (hunk 6 above)
Comment (vladikoff) [reply packages/fxa-content-server/app/scripts/templates/sign_in_totp_code.mustache]: in https://github.com/mozilla/fxa/issues/1297

File: packages/fxa-content-server/app/tests/spec/views/pair/auth_allow.js (hunk 7 above)
Comment (vladikoff) [reply packages/fxa-content-server/app/tests/spec/views/pair/auth_allow.js]: Yeah I had mixed feelings about this. I think we can drop this check for now, it doesn't add anything useful for users or the totp flow itself.

> Have you observed this happening in the wild

No we fixed that now.
